### PR TITLE
Salil/enso diags rebase

### PR DESCRIPTION
**Pull request by salilmahajan**
Porting ENSO atm diags to A-Prime infrastructure

### CODE REVIEW

**Comment by milenaveneziani** [comment]
Hi @salilmahajan: thanks for all the work on this.
I have rebased, but the rebase wasn't entirely straightforward: could you please check the result and see if you can recognize any possible problem in the files that you changed? (I have saved your original branch locally, just in case)
Most of the conflicts came out of the generate_html file, so that should definitely be checked. And then probably all the bash files that were *already* there and that you had to change to accommodate the changes in this PR.

In the meantime, I will test this on edison.

File: bash_scripts/aprime_atm_diags.bash [hunk 1]
```diff
@@ -340,16 +350,643 @@ while [ $j -lt $n_test_cases ]; do
    casename="${case_set[$j]}"
    scratch_dir="${scratch_dir_set[$j]}"
    begin_yr_ts=${begin_yr_ts_set[$j]}
+   end_yr_ts=${end_yr_ts_set[$j]}
 
    ./bash_scripts/plot_time_series.bash $scratch_dir \
 				        $casename \
 				        $begin_yr_ts \
+				        $end_yr_ts \
```
Comment (milenaveneziani) [inline bash_scripts/aprime_atm_diags.bash]: For example, I wonder why the changes in aprime_atm_diags up to this point are there in the first place. Seem like things unrelated to this PR.. Do you understand it?

File: bash_scripts/setup.bash [hunk 2]
```diff
@@ -20,16 +20,20 @@ if [ ! -d $test_scratch_dir_atm ]; then
...
 fi
 
-if [ "$ref_case" != "obs" ]; then
+#if [ "$ref_case" != "obs" ]; then
   ref_scratch_dir_atm=$ref_scratch_dir/atm
   if [ ! -d $ref_scratch_dir_atm ]; then
     mkdir $ref_scratch_dir_atm
   fi
-fi
+#fi
```
Comment (milenaveneziani) [inline bash_scripts/setup.bash]: we should remove the commented out lines at some point before merging.

File: bash_scripts/aprime_atm_diags.bash (hunk 1 above)
Comment (salilmahajan) [reply bash_scripts/aprime_atm_diags.bash]: Yes, I made these changes to better name log files and check for pre-existing files. I will look at the rebase now. Thanks @milenaveneziani .

**Comment by salilmahajan** [comment]
Hi Milena,


I just pulled my branch. How should I check if everything is okay. Did you mean that I go through the scripts to see discrepancies or did you have something else in mind.


Thanks,

Salil.


________________________________

**Comment by milenaveneziani** [comment]
Sorry I wasn’t clearer.
I just meant to go through some of the scripts, especially generate_html. I think that is OK, but just in case.
I went through them myself last night, and I have pointed out the places where I wasn’t sure (left comments on github).

Also, I just noticed that you changed the observation directory to point to a different one on your project space at NERSC and OLCF. Is it because you need additional observations that is not in the current obs_for_diagnostics directory? If so, I would suggest to put everything we need in $projdir/observations/Atm, so that we have all a-prime obs in one place (we typically have ocn/ice data in $projdir/observations/Ocean and $projdir/observations/SeaIce).
what do you think?

Milena
On Oct 13, 2017, at 10:11 AM, salilmahajan <notifications@github.com<mailto:notifications@github.com>> wrote:

Hi Milena,


I just pulled my branch. How should I check if everything is okay. Did you mean that I go through the scripts to see discrepancies or did you have something else in mind.


Thanks,

Salil.


________________________________

**Comment by milenaveneziani** [comment]
@salilmahajan: I have tested this on edison, and most things came out right. Various plots are missing though because I think the observations are missing. Could you please transfer the necessary data over to NERSC? (or maybe it's a problem with the obs location?).

**Comment by milenaveneziani** [comment]
my test can be seen here:
http://portal.nersc.gov/project/acme/milena/coupled_diagnostics_20170915.beta2.A_WCYCL1850S.ne30_oECv3_ICG.edison_years91-100_vs_obs/

File: bash_scripts/setup.bash (hunk 2 above)
Comment (salilmahajan) [reply bash_scripts/setup.bash]: yes, @milenaveneziani. I will do that. Regarding the scratch directory, I noticed earlier that there has been a slight modification in the path to the scratch space. I remember it being something like $projdir/$casename.scratch. But, now its $projdir/$plot_dir_name.scratch. $plots_dir_name is specific to the prescribed years. So, whenever we run the diags with new years, it would create a new directory and would now know that some files may already exist. Right? Should we change it to $projdir/$casename.scratch?

**Comment by salilmahajan** [comment]
@milenaveneziani: yes, it seems to be because of the data path. It now points to the location you suggested. Can you please try it one more time.

**Comment by milenaveneziani** [comment]
> Regarding the scratch directory, I noticed earlier that there has been a slight modification in the path to the scratch space. I remember it being something like $projdir/$casename.scratch. But, now its $projdir/$plot_dir_name.scratch. $plots_dir_name is specific to the prescribed years. So, whenever we run the diags with new years, it would create a new directory and would not know that some files may already exist. Right? Should we change it to $projdir/$casename.scratch?

yes, you are right.
The reason I changed it is that I started getting confused with the many different directories we got at the end of the analysis. One option could be to append coupled_diagnostics/$casename to $output_base_dir and then have the plots be in $output_base_dir/coupled_diagnostics/$casename/plots, the logs be in $output_base_dir/coupled_diagnostics/$casename/logs, etc. 
What do you think?

**Comment by salilmahajan** [comment]
Yes, we should reduce the number of directories generated. 

I am thinking we keep one scratch directory for each case where all generated data files would be saved:
$output_base_dir/coupled_diags_$casename.scratch

with each new diag run with different years getting its own directory with subdirectories for plots and log files:
$output_base_dir/coupled_diags_$casename_yearsX-Y_vs_$ref_case/coupled_diags_$casename_yearsX-Y_vs_$ref_case.plots
$output_base_dir/coupled_diags_$casename_yearsX-Y_vs_$ref_case/logs

I think we should keep the name of the case in the plots directory name to easily identify it when we transfer it to the $www_dir space.

**Comment by milenaveneziani** [comment]
That still seems like a lot of output directories to me (and not very contained as they all end up in output_base_dir), but I understand that it would be better to keep the casename in the plots directory.
Here is one more option, let me know if you like this better:
```
output_base_dir=$userChoice
....
plots_base_dir=$casename_yearsX-Y_vs_$ref_case
output_base_dir=$output_base_dir/coupled_diags_$casename
scratch_dir=$output_base_dir/scratch
plots_dir=$output_base_dir/$plots_base_dir/plots
logs_dir=$output_base_dir/$plots_base_dir/logs
www_dir=$www_dir/$plots_base_dir
```
This way we have one container for the casename case ($output_base_dir/coupled_diags_$casename) and the various diags cases within it, while all climos and other things go in the same scratch directory.

**Comment by salilmahajan** [comment]
Yes, this seems better. Thanks Milena.




________________________________

**Comment by salilmahajan** [comment]
@milenaveneziani, I just changed the directory structure on the branch. It works for the atm and enso diags on edison. Could you please check if it runs for the ocean diags too? Thanks!

File: python/plot_multiple_index.py [hunk 3]
```diff
@@ -0,0 +1,269 @@
...
+		ax[i].axis([plot_time[0],plot_time[-1], y_axis_ll, y_axis_ul])
+
+		print 'plot_time[0],plot_time[-1], 1.1*min_plot, 1.1*max_plot: ', \
+			plot_time[0],plot_time[-1], 1.1*min_plot, 1.1*max_plot
+
+		mean_line_plot = numpy.zeros(nt) + plot_ts_mean[i]
+		mean_line, = ax[i].plot(plot_time, mean_line_plot, color = 'black', linewidth = 1.0, label = 'Mean')
+
+		if begin_month == 0 and end_month == 11 and aggregate == 0:
+			bw   = 13
```
Comment (vanroekel) [inline python/plot_multiple_index.py]: @salilmahajan could you explain why a 1-yr running mean was chosen for the NINO3,4,3.4 indices?  Can this be changed to 5-months, which is the NINO index algorithm (http://www.cgd.ucar.edu/staff/trenbert/trenberth.papers/defnBAMS.pdf)

**Comment by salilmahajan** [comment]
@vanroekel, the choice of the bandwidth is only for visual purposes here, to smooth out seasonal variability. It could be changed to any appropriate number. I have seen other definitions too (e.g. https://www.ncdc.noaa.gov/teleconnections/enso/indicators/sst.php). The other computations and plots based on the indices (regressions) use the monthly index values in the diags; the smoothed values are not used anywhere.

**Comment by milenaveneziani** [comment]
Hi @salilmahajan: I am starting the testing of this again.

While I do that, I have a question regarding `test_begin_yr_enso_atm` and its 'end' correspondent: shouldn't they just be equal to, say `test_begin_yr_climateIndex_ts`, to make things consistent with the ElNino3.4 index calculated through MPAS-Analysis? (in that case, we would remove the `export test_end_yr_climateIndex_ts=9999` default we currently have, because that would not work with the atm scripts).

**Comment by milenaveneziani** [comment]
I am having problems with reading the observations in the new Atm directory on edison. @salilmahajan: can you please change the group to cli115?

**Comment by xylar** [comment]
@milenaveneziani, are you sure there's a group called cli115 on edison? Just want to make sure that's the right machine or group.

**Comment by milenaveneziani** [comment]
sorry, machine is right, group wrong :) I meant 'acme', not cli115.

**Comment by salilmahajan** [comment]
@milenaveneziani, I just changed the group settings.

**Comment by salilmahajan** [comment]
Yes @milenaveneziani, you are correct. The enso atm begin and end yr should be consistent with the climateindex_ts. I did not realize that before.

**Comment by milenaveneziani** [comment]
thanks @salilmahajan.

OK, I will make the begin_year and end_year change on my local branch and see how things go. You can then merge them into your branch if all tests pass.

**Comment by milenaveneziani** [comment]
@salilmahajan, here is the latest:
1) my test on edison passed almost completely this time. This results here:
http://portal.nersc.gov/project/acme/milena/20170915.beta2.A_WCYCL1850S.ne30_oECv3_ICG.edison_years96-100_vs_obs/
All the links work, except for the climatological global PRECP map, e.g. http://portal.nersc.gov/project/acme/milena/20170915.beta2.A_WCYCL1850S.ne30_oECv3_ICG.edison_years96-100_vs_obs/20170915.beta2.A_WCYCL1850S.ne30_oECv3_ICG.edison-GPCP_PRECT_climo_DJF.png
I looked at the corresponding log files, which you can find here:
`/global/project/projectdirs/acme/milena/coupled_diagnostics/20170915.beta2.A_WCYCL1850S.ne30_oECv3_ICG.edison_vs_obs/20170915.beta2.A_WCYCL1850S.ne30_oECv3_ICG.edison_years96-100_vs_obs.log/`
but could not understand the error this time. Could you please take a look?
2) the removing of begin_yr_enso_atm and end_yr_enso_atm seems to work well (see latest commit)
3) I also slightly adjusted the output directory structure. Let me know if you like what you see now.

After the edison test completes successfully, should I test on OLCF and aims4/acme1? Did you transfer the obs to the LLNL space2 common directory?

**Comment by salilmahajan** [comment]
@milenaveneziani, could you please open the permissions to your log files. Thanks!

Yes, I have tested the diags on Rhea, but not on aims yet. It will be great if you could test it on Rhea too.

**Comment by milenaveneziani** [comment]
permissions should be fixed.
Ok, I will test on titan, since you tested on rhea.
I can help with testing on aims/acme1 too: I should be able to just transfer the observations/Atm folder over to space2.

**Comment by milenaveneziani** [comment]
@salilmahajan: the test on titan can be seen here:
http://projects.olcf.ornl.gov/acme/milena/20161117.beta0.A_WCYCL1850S.ne30_oEC_ICG.edison_years31-40_vs_obs/
lots of the new nino diags plots are missing. Not sure why, permissions to the observations files seem OK.
Case directories are here:
`/lustre/atlas1/cli115/proj-shared/milena/aprime_titan/coupled_diagnostics`

Test on aims4 is underway.

**Comment by salilmahajan** [comment]
Yes @milenaveneziani, with the new commit the remapping on Nino time series variables is using a wrong variable switch.  I am fixing this now.

**Comment by milenaveneziani** [comment]
sorry about that.. But those plots were fine on edison though.

**Comment by salilmahajan** [comment]
I guess it would have used preexisting files that were generated from a prior diags run.

**Comment by milenaveneziani** [comment]
yeah, that's exactly what I was about to comment.

**Comment by salilmahajan** [comment]
Hi @milenaveneziani, this should fix the PRECT climo problem too.

**Comment by milenaveneziani** [comment]
ok, I have purged the atm plots and scratch directory and restarted all tests.

**Comment by milenaveneziani** [comment]
@salilmahajan: looks like the new commit introduced a new error in the Meridional tropical averages, for TS and TAUX only (PRECP seems OK). Check this log file on titan, for example:
`/ccs/home/milena/proj_cli115/milena/aprime_titan/coupled_diagnostics/20161117.beta0.A_WCYCL1850S.ne30_oEC_ICG.edison_vs_obs/20161117.beta0.A_WCYCL1850S.ne30_oEC_ICG.edison_years31-40_vs_obs.logs/plot_climo_20161117.beta0.A_WCYCL1850S.ne30_oEC_ICG.edison-ERS.TAU.ANN.log`

Good news is previous problems are solved.

**Comment by salilmahajan** [comment]
@milenaveneziani, it should be fixed now. Thanks!

**Comment by milenaveneziani** [comment]
Tests passed on titan!

Re-resting on edison and aims4 now.

**Comment by milenaveneziani** [comment]
Test on aims4 passed: I can see all plots now over there too.
Same goes with edison.

I will now do a final test on acme1 and then merge if all goes well.

**Comment by milenaveneziani** [comment]
All tests passed.

One thing I just realized: the new python files (and maybe others?) are missing the license headers. @kevans32: do you run a script to do that? 
Anyways, I am going to update MPAS-Analysis in a-prime very soon, so we can add the headers at that point.
